IPv4 RFC791 fragmentation

aperezdc opened the pull request:
Same as [PR#26](https://github.com/Igalia/snabbswitch/pull/26), with most of the things from the code review fixed. The only missing bit is some unit tests, which I will try to add.

kbara commented:
https://github.com/SnabbCo/snabbswitch/blob/7ec1ae2e06e93e70a39cfc3149fae25a67992312/src/apps/lwaftr/lwutil.lua now exists; it has some of the utility functions you've been wishing for, for reading/writing 16/32 bit values, and for getting IHLs.

dpino commented on src/apps/lwaftr/fragmentv4.lua:
```diff
@@ -0,0 +1,105 @@
+local fresh_frag_id = (function ()
+   local internal_frag_id = 0x4242
+   return function ()
+      internal_frag_id = (internal_frag_id + 1) % 0xFFFF
```
Forget about this one. Modulo 0xFFFF will only give values in the 0-0xFFFE range, 65535 values, and not 65536 values, which is what it was intended.

In addition to that, calculating the reminder of a division with bitwise and arithmetic is usually faster than using mod (I have no numbers for LuaJIT, but is a well-known optimization). So, in conclusion, I think the code is fine as it originally was. Sorry for the noise.

aperezdc replied:
No problem, I'm reverting back to using `band(...)`, then.

aperezdc commented:
PTAL. Changes in the latest update:

* Changed back modulus operation to bitwise-and to ensure values in 16-bit ranges for `fresh_frag_id()`.
* Added a `selftest()` function. This can be executed manually using: `sudo ./snabb snsh -l apps.lwaftr.fragmentv4 -e 'apps.lwaftr.fragmentv4.selftest()'`
* Corrected a couple of mistakes found after writing the `selftest()` function.

aperezdc commented:
Updated again after talking with @kbara and agreeing on returning a status code (`FRAGMENT_OK`, `FRAGMENT_UNNEEDED`, or `FRAGMENT_FORBIDDEN`) along with the return value.

dpino commented on src/apps/lwaftr/fragmentv4.lua:
```diff
@@ -0,0 +1,289 @@
+
+   -- Discard packets with the DF (dont't fragment) flag set
+   do
+		local flags_and_frag_offset = C.ntohs(rd16(ipv4_pkt.data + flags_and_frag_offset_offset))
```
I'm afraid some tabs slipped in. SnS uses 3-space indents for Lua code, no tabs. I have the git pre-commit hook activated, so it prevents me from commiting tabs, trailing white spaces, etc in the source code.

aperezdc replied:
Ack.

dpino commented on src/apps/lwaftr/fragmentv4.lua:
```diff
@@ -0,0 +1,289 @@
+   code, result = assert(fragment_ipv4(pkt, 1000))
+   assert(code == FRAGMENT_OK)
+   assert(#result == 2,
+          "fragmentation returned " .. tostring(#result) .. " packets (2 expected)")
```
Concatenating a number to a string automatically converts the number to a string, no need to explicitly coerce it.

aperezdc replied:
Ack.

dpino commented on src/apps/lwaftr/fragmentv4.lua:
```diff
@@ -0,0 +1,289 @@
+
+   -- Makes an IPv4 packet, with Ethernet framing, with a given payload size
+   local function make_ipv4_packet(payload_size)
+      local pkt = packet.allocate()
```
It's also possible to create a packet out of raw data. There's an example of this in ```lib/protocol/tcp.lua```:

```lua
   local p = packet.from_string(lib.hexundump([[
      52:54:00:02:02:02 52:54:00:01:01:01 08 00 45 00
      00 34 59 1a 40 00 40 06 00 00 c0 a8 14 a9 6b 15
      f0 b4 de 0b 01 bb e7 db 57 bc 91 cd 18 32 80 10
      05 9f 00 00 00 00 01 01 08 0a 06 0c 5c bd fa 4a
      e1 65
   ]], 66))
```
Maybe it was possible to use something like that here. The code as it is now, it's OK to me.

dpino commented:
LGTM.

aperezdc commented:
Updated the indentation and string concatenation nits.

The packet creation seems clearer to me by calling methods in an object than using `hexundump()`, and tests do not need to be super fast or anything, so as long as you don't mind I would rather keep the `make_ipv4_packet()` function as it is now.

dpino commented:
@aperezdc Fine for me to use objects rather than `hexundump()`, it was more like a JFYI note :) Maybe it can be handy next time.

LGTM. I'd merge it but I prefer @kbara confirms before doing it.

aperezdc commented:
Added calculation of IP checksums for generated packets, plus one extra testcase which reassembles a packet and verifies that the pattern of bytes in the payload is the same as the one generated as input.

aperezdc commented:
All seems to be working properly now: running `end-to-end.sh`, and inspecting both the input packet and the output packets, the TCP payload reassembled by Wireshark is the same, byte-by-byte, to the one from `apps/lwaftr/data/tcp-ipv6-fromb4-toinet-1046.pcap`.

aperezdc commented:
Updated with latest version of @kbara's integration patch and tests. Now `end-to-end.sh` reports all the tests as successfully passed :smile:

kbara commented:
I'll merge it, but a patch to separate the unit tests into their own file would still be very welcome. Thanks. :-)
